utils/utils.py

```diff
@@ -0,0 +1,299 @@
+import csv
+import os
+import pynini
+from pynini.lib import pynutil,utf8
+import string
+
+def get_abs_path(rel_path):
+    """
+    Get absolute path
+    Args:
+        rel_path: relative path to this file
+        
+    Returns absolute path
+    """
+    return os.path.dirname(os.path.abspath(__file__)) + '/' + rel_path
+
+
+def load_labels(abs_path):
+    """
+    loads relative path file as dictionary
+    Args:
+        abs_path: absolute path
+    Returns dictionary of mappings
+    """
+    label_tsv = open(abs_path, encoding="utf-8")
+    labels = list(csv.reader(label_tsv, delimiter="\t"))
+    return labels
+
+
+DEFAULT_INSERT_COST = 1.
+DEFAULT_DELETE_COST = 1.
+DEFAULT_SUBSTITUTE_COST = 1.
+
+
+class Error(Exception):
+  """Errors specific to this module."""
+
+  pass
+
+
+class EditTransducer:
+  """Factored edit transducer.
+  This class stores the two factors of an finite-alphabet edit transducer and
+  supports insertion, deletion, and substitution operations with user-specified
+  costs.
+  Note that the cost of substitution must be less than the cost of insertion
+  plus the cost of deletion or no optimal path will include substitution.
+  One can impose an upper bound on the number of permissible edits by
+  setting a non-zero value for `bound`. This often results in substantial
+  improvements in performance.
+  """
+
+  # Reserved labels for edit operations.
+  DELETE = "<delete>"
+  INSERT = "<insert>"
+  SUBSTITUTE = "<substitute>"
+
+  def __init__(self,
```

Comment:
code reorg:

EditTransducer should be constructed from a token list (or sometimes called *alphabet* in char-based system, *lexicon* in word-based system), sentence_A & sentence_B processing code should move outside of EditTransducer class